server/filestore.go

```diff
@@ -3984,6 +4005,19 @@ func (mb *msgBlock) setupWriteCache(buf []byte) {
 	mb.startCacheExpireTimer()
 }
 
+// finishedWithCache removes the strong reference to the cache, instead
+// returning it to the elastic pointer. Note that the elastic pointer can
+// be either weak or strong, they are strengthened when there are pending
+// writes and weak otherwise.
+func (mb *msgBlock) finishedWithCache() {
+	if mb.cache != nil {
+		// The reference mb.cache shouldn't have changed ordinarily, but
+		// stash it away into the weak pointer again just in case.
+		mb.ecache.Set(mb.cache)
+		mb.cache = nil
```

Comment:
We have two levels in the cache, the indexing and the data itself. Both IIRC are under mb.cache but they have different lifetimes under normal operations. This seems to undo that?

Reply:
They're still held in the `ecache` until it is GC'd (and for that there would need to be no strong references left too), but if that happens, we'll just reload the block and re-index it the next time we need to.

Reply:
fss is held outside this scope. Not sure if we want this extended to that in a follow on PR or not.